### DISCUSSION TITLE
Use Pocketbase query syntax directly in framework code

### DISCUSSION THREAD

**Discussion by zakkor** [pocketbase/pocketbase#2309] [Q&A] [answered]
The Pocketbase query syntax is very convenient & makes complex queries very easy to do.

As far as I can tell, this can only be used in client code, and not in framework code, unless I missed something?

For example, something like:

`app.Dao().Query("user.assocs ~ \"yi67v7dmpb38cby\"")` (just a random example)

instead of the current way of doing queries in the framework, which would require building the query manually.

Since this is possible to do in the web UI and in client code, Pocketbase *should* be able to do this. Is there any way to do it?

**Comment by ganigeorgiev** [comment, marked as answer]
There is no exposed direct helper for this at the moment since for the records it works with the request data and is a little more complicated. Here is a simplified helper that you can try (_I didn't tested it_):

```go
func searchRecords(app core.App, baseCollectionNameOrId string, filter string) ([]*models.Record, error) {
	collection, err := app.Dao().FindCollectionByNameOrId(baseCollectionNameOrId)
	if err != nil {
		return nil, err
	}

	// construct a base query
	q := app.Dao().RecordQuery(collection)

	// build a fields resolver and attach the generated conditions to the query
	// ---
	resolver := resolvers.NewRecordFieldResolver(
		app.Dao(),
		collection, // the main/base collection
		nil,        // pass a request data pointer if you want to use `@request.*` filters
		true,       // allow searching hidden/protected fields like "email"
	)

	expr, err := search.FilterData(filter).BuildExpr(resolver)
	if err != nil {
		return nil, err
	}

	resolver.UpdateQuery(q) // attaches any adhoc joins and aliases
	q.AndWhere(expr)        // attach the conditions generated from the filter
	// ---

	records := []*models.Record{}

	if err := q.All(&records); err != nil {
		return nil, err
	}

	return records, nil
}

records, err := searchRecords(app, "example", `user.assocs.id ?= "SOME_ID"`)
```

_(I'll consider exposing something like the above eventually in the public apis sometime in the future)_

**Reply by zakkor** [reply]
Tested, that works perfectly! Thanks

**Reply by ganigeorgiev** [reply]
_Just for info in case someone stumble here - `app.Dao().FindRecordsByFilter()` and `app.Dao().FindFirstRecordByFilter()` were added in the `develop` branch and will be available with the next v0.17.0 release (there are no ETAs yet)._

**Reply by hungcrush** [reply]
@ganigeorgiev Very nice~~ thank you so much
